### ISSUE TITLE
Simple transfer Learning detect Ship exist, U-Net base on ResNet34 Transfer learning, Unet34 + Predict result

### ISSUE THREAD

**Issue by baekspace88** [KGGS/Airbus-Ship-Detection-Challenge#3]
Simple transfer Learning detect Ship exist (Keras)
U-Net base on ResNet34 Transfer learning (Keras)
Unet34 + Predict result

https://www.kaggle.com/super13579/simple-transfer-learning-detect-ship-exist-keras
https://www.kaggle.com/super13579/u-net-base-on-resnet34-transfer-learning-keras
https://www.kaggle.com/super13579/unet34-predict-result

**Comment by baekspace88** [comment]
### ResNet
- https://datascienceschool.net/view-notebook/958022040c544257aa7ba88643d6c032/
- https://dnddnjs.github.io/cifar10/2018/10/09/resnet/
- https://ratsgo.github.io/deep%20learning/2017/10/09/CNNs/
- 2015년 ILSVRC 대회에서 우승한 모델
- Degradation 문제 : 모델의 깊이가 깊어지다 보면 어느 순간 더 얕은 모델의 성능보다 더 나빠지는 현상
- <img width="513" alt="스크린샷 2020-02-01 오후 6 56 19" src="https://user-images.githubusercontent.com/48514061/73590327-90480680-4524-11ea-9b26-b0017405f1c4.png">
- Skip connection : 네트워크의 입력과 출력이 더해진 것을 다음 레이어의 입력으로 사용
- Residual learning : 입력과 출력의 잔차를  𝐹(𝑥)=𝐻(𝑥)−𝑥 라고 정의. 네트워크는 이  𝐹(𝑥)(잔차) 를 0이 되도록 학습
- <img width="478" alt="스크린샷 2020-02-01 오후 6 38 03" src="https://user-images.githubusercontent.com/48514061/73590131-0860fd00-4522-11ea-9572-b6ce7e6d84df.png">
- 덧셈 연산의 추가. 이는 추가적인 연산량이나 파라미터가 많이 필요하지 않음. 또한 역전파 시에 그레디언트가 잘 흘러갈 수 있게 해준다는 장점
- Pre-activation Residual Unit : skip connection 위치를 바꿈으로써 다음 레이어의 입력 값에 이전 레이어의 입력을 그대로 반영하고 각 레이어간의 관계가 더하기로 표현 되기 때문에 수학적으로 좀 더 유용
- <img width="496" alt="스크린샷 2020-02-01 오후 6 46 26" src="https://user-images.githubusercontent.com/48514061/73590212-2e3ad180-4523-11ea-87d1-2b36c7ef31d4.png">
- backbone으로의 역할 : ResNet과 같은 classification model은 입력을 받아서 각 task에 맞는 모듈로 전달해주는 역할. 객체를 검출하든 영역들을 나누든 Neural Network는 입력 이미지로부터 다양한 feature를 추출해주기 때문
- Residual block이 앙상블(ensemble) 모델을 구축한 것과 비슷한 효과를 냄. (n개 skip connection이 있다면 2^n개의 서로 다른 통로 존재)
- <img width="662" alt="스크린샷 2020-02-01 오후 6 58 53" src="https://user-images.githubusercontent.com/48514061/73590361-e87f0880-4524-11ea-8feb-50f3875c3fb7.png">



### U-Net
- https://towardsdatascience.com/introduction-to-u-net-and-res-net-for-image-segmentation-9afcb432ee2f
- https://modulabs-biomedical.github.io/U_Net
- https://mylifemystudy.tistory.com/87
- <img width="655" alt="스크린샷 2020-02-01 오후 7 33 58" src="https://user-images.githubusercontent.com/48514061/73590794-e79ca580-4529-11ea-8a66-6a180776f1db.png">
- Architecture
   - Convolution Encoder에 해당하는 Contracting Path + Convolution Decoder에 해당하는 Expanding Path의 구조로 구성
   - Expanding Path에서 Upsampling 할 때, 좀 더 정확한 Localization을 하기 위해서 Contracting Path의 Feature를 Copy and Crop하여 Concat 하는 구조

<img width="642" alt="스크린샷 2020-02-01 오후 7 34 15" src="https://user-images.githubusercontent.com/48514061/73590798-f5eac180-4529-11ea-9e48-df411ae4cb82.png">

- 특징 : 
   - Contracting Path에서 Pooling되기 전의 Feature들은 Upsampling 시에 Layer와 결합되어 고해상도 output을 만들어 낼 수 있음
   - Overlap-tile strategy : 
     - Blue tile: Input, Yellow tile: Output (contracting path에서 padding이 없었기 때문에 점점 이미지 외곽 부분이 없어진 결과) -> blue area에 기반하여 yellow area의 segmentation을 예측
     - Mirroring 방법으로 pixel의 값을 채워줌.(zero-padding이 아닌 mirror padding)
     - 이미 검증이 끝난 곳은 아예 건너 뛰고, 다음 patch부터 새 검증 (overlap 비율이 낮음) -> 빠른 속도
   - context 인식과 localization의 trade-off에 빠지지 않음.
      - Contracting path : 이미지의 context를 포착
      - Expansive path : 피쳐맵을 Upsampling하고 이를 Contracting path에서 포착한 피쳐맵의 context와 결합하여 더 정확한 localization을 함.
   - Data Augmentation
     - Elastic Deformation
     - Weighted Cross Entropy + Data Augmentation
        Morphological 연산을 사용해 각 instance의 테두리와 테두리 사이가 반드시 배경이 되도록 하는 처리
<img width="843" alt="스크린샷 2020-02-08 오후 11 49 16" src="https://user-images.githubusercontent.com/48514061/74087229-a5490a80-4acd-11ea-976d-fc31e17f5146.png">

**Comment by baekspace88** [comment]
- RLE 압축 : https://ko.wikipedia.org/wiki/%EB%9F%B0_%EB%A0%9D%EC%8A%A4_%EB%B6%80%ED%98%B8%ED%99%94

**Comment by baekspace88** [comment]
- train_test_split
   stratify : 지정한 Data의 비율을 유지한다. 예를 들어, Label Set인 Y가 25%의 0과 75%의 1로 이루어진 Binary Set일 때, stratify=Y로 설정하면 나누어진 데이터셋들도 0과 1을 각각 25%, 75%로 유지한 채 분할된다.
- pandas.DataFrame.clip
   Trim values at input threshold(s).

**Comment by baekspace88** [comment]
#### UpSampling2D Layer vs. Conv2DTranspose Layer
- https://machinelearningmastery.com/upsampling-and-transpose-convolution-layers-for-generative-adversarial-networks/
- https://zzsza.github.io/data/2018/06/25/upsampling-with-transposed-convolution/
- UpSampling Layer
   - The layer has no parameters or model weights. It is just doubling the input.
~~~
# example of using the upsampling layer
from numpy import asarray
from keras.models import Sequential
from keras.layers import UpSampling2D
# define input data
X = asarray([[1, 2],
			 [3, 4]])
# show input data for context
print(X)
# reshape input data into one sample a sample with a channel
X = X.reshape((1, 2, 2, 1))
# define model
model = Sequential()
model.add(UpSampling2D(input_shape=(2, 2, 1)))
# summarize the model
model.summary()
# make a prediction with the model
yhat = model.predict(X)
# reshape output to remove channel to make printing easier
yhat = yhat.reshape((4, 4))
# summarize output
print(yhat)
~~~
<img width="625" alt="스크린샷 2020-02-09 오전 12 37 10" src="https://user-images.githubusercontent.com/48514061/74088152-5a7fc080-4ad6-11ea-8039-2892b7fdcdbd.png">

- Transpose Convolutional Layer
   - The layer that combines the UpSampling2D and Conv2D layers into one layer
   - Convolution된 것을 다시 패턴을 유지하며 UpSampling하고자 함.
   - 1X1 filter 1개, 2X2 stride 에 의해 파라미터와 새로운 row, column이 추가되었음. (학습 가능)
   - 가중치들은 학습 가능하기 때문에 미리 정의된 보간 방법이 필요하지 않음.
   - filter의 사이즈와 개수를 조정하면 random weight로 초기화
~~~
from numpy import asarray
from keras.models import Sequential
from keras.layers import Conv2DTranspose
# define input data
X = asarray([[1, 2],
			 [3, 4]])
# show input data for context
print(X)
# reshape input data into one sample a sample with a channel
X = X.reshape((1, 2, 2, 1))
# define model
model = Sequential()
model.add(Conv2DTranspose(1, (1,1), strides=(2,2), input_shape=(2, 2, 1)))
# summarize the model
model.summary()
# define weights that they do nothing
weights = [asarray([[[[1]]]]), asarray([0])]
# store the weights in the model
model.set_weights(weights)
# make a prediction with the model
yhat = model.predict(X)
# reshape output to remove channel to make printing easier
yhat = yhat.reshape((4, 4))
# summarize output
print(yhat)
~~~
<img width="624" alt="스크린샷 2020-02-09 오전 12 50 48" src="https://user-images.githubusercontent.com/48514061/74088139-45a32d00-4ad6-11ea-9d9d-a1aa1e8d3498.png">
